packages/node/src/lib/token-manager.ts

```diff
@@ -0,0 +1,290 @@
+import { uuid } from './uuid'
+import {
+  FetchHTTPClient,
+  HTTPClient,
+  HTTPClientRequest,
+  HTTPResponse,
+} from './http-client'
+import { SignOptions, sign } from 'jsonwebtoken'
+import { Emitter, backoff, sleep } from '@segment/analytics-core'
+import { AbortSignal, AbortController } from './abort'
+
+type AccessToken = {
+  access_token: string
+  expires_in: number
+}
+
+export type OAuthSettings = {
+  clientId: string
+  clientKey: Buffer
+  keyId: string
+  authServer?: string
+  scope?: string
+  maxRetries?: number
+  httpClient?: HTTPClient
+}
+
+export class TokenManager {
+  private alg = 'RS256' as const
+  private grantType = 'client_credentials' as const
+  private clientAssertionType =
+    'urn:ietf:params:oauth:client-assertion-type:jwt-bearer' as const
+  private clientId: string
+  private clientKey: Buffer
+  private keyId: string
+  private scope: string
+  private authServer: string
+  private httpClient: HTTPClient
+  private maxRetries: number
+  private clockSkewInSeconds = 0
+
+  private controller: AbortController
+  private signal: AbortSignal
+
+  private accessToken?: AccessToken
+  private tokenEmitter = new Emitter<{
+    access_token: [{ token: AccessToken } | { error: unknown }]
+  }>()
+  private retryCount: number
+  private lastError: any
+  private pollerTimer?: ReturnType<typeof setTimeout>
+
+  constructor(props: OAuthSettings) {
+    this.keyId = props.keyId
+    this.clientId = props.clientId
+    this.clientKey = props.clientKey
+    this.authServer = props.authServer ?? 'https://oauth2.segment.io'
+    this.scope = props.scope ?? 'tracking_api:write'
+    this.httpClient = props.httpClient ?? new FetchHTTPClient()
+    this.maxRetries = props.maxRetries ?? 3
+    this.tokenEmitter.on('access_token', (event) => {
+      if ('token' in event) {
+        this.accessToken = event.token
+      }
+    })
+    this.controller = new AbortController()
+    this.signal = this.controller.signal
+    this.signal.addEventListener('abort', () => {
+      clearTimeout(this.pollerTimer)
+    })
+    this.retryCount = 0
+  }
+
+  async pollerLoop() {
+    let timeUntilRefreshInMs = 25
+    let response: HTTPResponse
+
+    try {
+      response = await this.requestAccessToken()
+    } catch (err) {
+      // Error without a status code - likely networking, retry
+      this.retryCount++
+      this.lastError = err
+
+      if (this.retryCount % this.maxRetries == 0) {
```

Comment:
Ah, so this doesn't actually stop the poller, just defines the frequency we log errors?

Reply:
It defines the number retries before it emits an error, releasing any waiting get_token requests. ... but now that I think about it the logic is disconnected because the actual token fetches will be happening in the background.  Any get_token may wait seconds or error out instantly based on chance.  Open to suggestions, probably would be better to stop it and let a new request start it up again.  Don't need to spam a server with bad requests, but also we won't be able to do anything without a token.  We depend on the developer to respond to errors, really.